### ISSUE TITLE
Kodi sends a request for /content/media/object_id/<object-id>/res_id/<res-id>/group/default/ext/ in order to scan external subtitle files

### ISSUE THREAD

**Issue by masnagam** [gerbera/gerbera#2841]
A recent Kodi (I confirmed this problem only with Kodi/Nexus) tries to scan external subtitle files for a video file by default.

Kodi's debug logs are shown below:

```
2023-07-02 12:56:46.553 T:82      debug <general>: CCurlFile::GetMimeType - <http://$HOST:49494/content/media/object_id/3748/res_id/0/group/default/ext/file.mpeg> -> video/mpeg
2023-07-02 12:56:46.553 T:82       info <general>: Creating InputStream
2023-07-02 12:56:46.553 T:82      debug <general>: CurlFile::Open - <http://$HOST:49494/content/media/object_id/3748/res_id/0/group/default/ext/file.mpeg>
2023-07-02 12:56:46.554 T:32      debug <general>: Loading settings for upnp://$UUID/3748/
2023-07-02 12:56:46.569 T:7       debug <general>: CGLContextEGL::SwapBuffers: sync check blocking
2023-07-02 12:56:46.585 T:7       debug <general>: CGLContextEGL::SwapBuffers: sync sleep: 16368
2023-07-02 12:56:46.619 T:82    warning <general>: CCurlFile::Open - <http://$HOST:49494/content/media/object_id/3748/res_id/0/group/default/ext/file.mpeg> Disabling multi session due to broken libupnp server
2023-07-02 12:56:46.619 T:82      debug <general>: CFileCache::Open - <http://$HOST:49494/content/media/object_id/3748/res_id/0/group/default/ext/file.mpeg> opening
2023-07-02 12:56:46.619 T:82      debug <general>: CurlFile::Open - <http://$HOST:49494/content/media/object_id/3748/res_id/0/group/default/ext/file.mpeg>
2023-07-02 12:56:46.642 T:82    warning <general>: CCurlFile::Open - <http://$HOST:49494/content/media/object_id/3748/res_id/0/group/default/ext/file.mpeg> Disabling multi session due to broken libupnp server
2023-07-02 12:56:46.642 T:82      debug <general>: CFileCache::Open - <http://$HOST:49494/content/media/object_id/3748/res_id/0/group/default/ext/file.mpeg> source chunk size is 0, setting cache chunk size to 131072
2023-07-02 12:56:46.642 T:82      debug <general>: CFileCache::Open - <http://$HOST:49494/content/media/object_id/3748/res_id/0/group/default/ext/file.mpeg> using single memory cache sized 20971520 bytes
2023-07-02 12:56:46.642 T:84      debug <general>: Thread FileCache start, auto delete: false
2023-07-02 12:56:46.642 T:82      debug <general>: ScanForExternalSubtitles: Searching for subtitles...
2023-07-02 12:56:46.642 T:82      debug <general>: CurlFile::Open - <http://$HOST:49494/content/media/object_id/3748/res_id/0/group/default/ext/>
2023-07-02 12:56:46.642 T:82      debug <general>: easy_acquire - Created session to http://$HOST
2023-07-02 12:56:46.705 T:82    warning <general>: CCurlFile::Open - <http://$HOST:49494/content/media/object_id/3748/res_id/0/group/default/ext/> Disabling multi session due to broken libupnp server
```

As you can see, Kodi sends a request for `/content/media/object_id/3748/res_id/0/group/default/ext/` in order to scan entries in the parent folder.

I confirmed that Gebera returns the video file for the following URLs

Gerbera returns the video file when sending requests for the following URLs:

* content/media/object_id/3748/res_id/0/group/default/ext/file.mpeg
* content/media/object_id/3748/res_id/0/group/default/ext/
* content/media/object_id/3748/res_id/0/group/default/
* content/media/object_id/3748/res_id/0/group/
* content/media/object_id/3748/res_id/0/
* content/media/object_id/3748/res_id/
* content/media/object_id/3748/

As a result, the video playback gets stuck until the entire video file is loaded (but it's not used for the playback, used for scanning external subtitle files).

I explored the GitHub issues and now I've known that this Gerbera's behavior is intentional, but this behavior may cause other similar issues.

Is there any workaround or recommended configurartion for Gerbera in order to avoid this situation?
Or we have to modify Gerbera and check the request URL before creating `FileRequestHandler` or `FileIOHandler`?

**Comment by KarlStraussberger** [comment]
I expect you're running gerbera 1.12.1

The upnp behaviour is sending subtitle files as resources. I tested our subtitle also with kodi (when we last changed it). Can you also cut a log from gerbera to see the response.

**Comment by masnagam** [comment]
Thank you for your quick reply.

> I expect you're running gerbera 1.12.1

Yes. I'm using the latest version (1.12.1).

> Can you also cut a log from gerbera to see the response.

OK, I'll do.  Wait a moment.

I saw this problem using https://github.com/mirakc/docker-timeshift-x/tree/main/gerbera.
Video files are imported by `import.js`.
`config.xml` is created in `run.sh`.

**Comment by masnagam** [comment]
I added a Gerbera server to Kodi as a UPnP device by using the "Add videos..." menu.
I played a video in the UPnP device and then got [log.txt](https://github.com/gerbera/gerbera/files/11933786/log.txt).
The video has no external subtitle file.

Kodi opened a video file for playback:

> 2023-07-03 15:25:06.602  trace: [file_request_handler.cc:207] open(): start: /content/media/object_id/3770/res_id/0/group/default/ext/file.mpeg

Kodi opened the parent folder in order to scan external subtitle files:

> 2023-07-03 15:25:06.661  trace: [file_request_handler.cc:207] open(): start: /content/media/object_id/3770/res_id/0/group/default/ext/

Kodi got stuck about 2 minutes for loading the entier video data for the scan:

> 2023-07-03 15:25:06.692  trace: [content_manager.cc:1921] triggerPlayHook(): Marking object 64A21D90.ニュース・気象情報[字] as played
>2023-07-03 15:25:06.693  trace: [content_manager.cc:1930] triggerPlayHook(): end
>2023-07-03 15:27:17.156  trace: [server.cc:543] createRequestHandler(): Filename: /content/media/object_id/3770/res_id/0/group/default/ext/file.mpeg

Related methods in Kodi:

* https://github.com/xbmc/xbmc/blob/dfc1dafbe6097af1f8a596309826fe15fee36c71/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L798
* https://github.com/xbmc/xbmc/blob/dfc1dafbe6097af1f8a596309826fe15fee36c71/xbmc/Util.cpp#L1940.

**Comment by KarlStraussberger** [comment]
We cannot do anything about the implementation of Kodi to expect that the items URI can be handled like a folder path, but we can configure gerbera to search for subtitles itself and add the respective <res .../> element. After that Kodi should not do its own magic.

You should add the following in the `<import>` element adapt to your needs and rescan your media afterwards
```
		<resources case-sensitive="false">
			<container location="image" parentCount="2" minDepth="2">
				<add-file name="%album%.jpg"/>
				<add-file name="%filename%.jpg"/>
				<add-file name="%title%.jpg"/>
				<add-file name="Folder.jpg"/>
				<add-file name="cover.jpg"/>
			</container>
			<fanart>
				<add-file name="%album%.jpg"/>
				<add-file name="%filename%-poster.jpg"/>
				<add-file name="%filename%-landscape.jpg"/>
				<add-file name="%filename%-fanart.jpg"/>
				<add-file name="cover.png"/>
				<add-file name="cover.jpg"/>
				<add-file name="album.jpg"/>
				<add-file name="front.jpg"/>
			</fanart>
			<subtitle>
				<add-file name="%filename%.srt"/>
				<add-file name="%filename%.Deutsch.srt"/>
				<add-dir name="%filename%" ext="srt"/>
			</subtitle>
			<metafile>
				<add-file name="%filename%.nfo"/>
			</metafile>
			<resource>
				<add-file name="%filename%.srt"/>
			</resource>
		</resources>
```

**Comment by masnagam** [comment]
I succeeded to add several "dummy" resources to a video object by adding the following `<resources>` element to `<import>`:

```xml
  <import hidden-files="no">
    <scripting script-charset="UTF-8">
      <virtual-layout type="js">
        <import-script>/usr/local/share/gerbera/js/import.js</import-script>
      </virtual-layout>
    </scripting>
    ...
    <resources case-sensitive="false">
      <subtitle>
        <add-file name="/run.sh" />
        <add-dir name="/media/timeshift" ext="json" />
      </subtitle>
    </resources>
  </import>
```

The `<add-file>` above didn't work, but the `<add-dir>` worked and I confirmed several json files were added to a video object from the Web UI by clicking the "edit" button.

I confirmed that each resource has its own **unique ID**.  For example,

* main resource (video file)
  * /content/media/object_id/2258/res_id/**0**/ext/file.mpeg
* first dummy resource (json file)
  * /content/media/object_id/2258/res_id/**1**/type/json/ext/file.nhk.timeshift.json

As a result, `curl http://$HOST:$PORT/content/media/object_id/2258/res_id/0/ext/` still returns the entire video file.
So, we cannot solve this issue by adding a `<resources>` element.

For solving this issue in Gerbera side, Gerbera needs to handle `/content/media/object_id/2258/res_id/0/ext/` as a folder, but it's not supported at this point (if my understanding is correct).  (Of cource, we can solve this issue by modifying Kodi and disabling scanning external subtitle files)

In the first place, I wonder if a filename like `/content/media/object_id/2258/res_id/0/ext/` is a valid URL in Gerbera.  `URLUtils::parseParameters()` can parse such a filename, but the `ext` is parsed as a parameter name and not as a folder name.  As a result the `ext` parameter has no value.  Is this intentional?  If such a filename is invalid, I think it might be better that Gerbera rejects the request.

**Comment by KarlStraussberger** [comment]
I don't know who came up with the idea that the URL reflects a folder structure.

In UPnP you have browse requests that return the children of a given (parent) object. Each object returns related information like thumbnails and subtitles as `<res>` which may point to streams in the media file or additional files on the disk somewhere. Your URL is structured like REST calls are and refers to resource number 0 of object 2258 (which is the media file in most cases). Anything after that is handled as parameters when streaming the resource to the client.

In addition to that, gerbera supports virtual structures where the path to a media file is not the physical location but based on metadata like artist or year. Each virtual item has its own id but refers to the same resources.

**Comment by masnagam** [comment]
OK, I understand.
I close this ticket.
Thank you for your kindly explanation :)

**Comment by masnagam** [comment]
this issue was solved in xbmc/xbmc@11331488382c6f3ea450c53a0decc32484193c29.
Kodi 21.1 works fine now.

**Comment by KarlStraussberger** [comment]
Thank you for reporting back
